### DISCUSSION TITLE
Avoid Suspending Atom with Async Default After Initial Fetch

### DISCUSSION THREAD

**Discussion by stevemolitor** [pmndrs/jotai#2144] [Q&A] [answered]
See this [CodeSandbox](https://codesandbox.io/s/frosty-shadow-sjg4qn?file=/src/App.tsx). It's suspending when I don't want it to. My intent is to suspend while fetching the initial version of a resource, and then let the user edit in memory without suspending.

I have an `iceCream` atom that I initialize with a promise:

```ts
const iceCreamDb: IceCream = { flavor: "chocolate" };

const fetchIceCream = async (): Promise<IceCream> =>
  Promise.resolve(iceCreamDb);

const iceCreamAtom = atomWithDefault<Promise<IceCream> | IceCream>(
  fetchIceCream
);
```

I have another atom that just grabs the "flavor" from the ice cream atom:

```ts
const flavorAtom = atom(
  async (get) => {
    const { flavor } = await get(iceCreamAtom);
    return flavor;
  },
  (get, set, flavor: Flavor) => {
    const iceCream = get(iceCreamAtom) as IceCream;
    set(iceCreamAtom, {
      ...iceCream,
      flavor,
    });
  }
);
```

I have a `FlavorPicker` component that uses `flavorAtom` to render a flavor picker:

```ts
const FlavorPicker = () => {
  const [flavor, setFlavor] = useAtom(flavorAtom);

  const onSelect = useCallback<ChangeEventHandler<HTMLSelectElement>>(
    (event) => {
      setFlavor(event.target.value as Flavor);
    },
    [setFlavor]
  );

  return (
    <label>
      Flavor:{" "}
      <select name="flavor" value={flavor} onChange={onSelect}>
        <option value="none" disabled>
          Pick Flavor:
        </option>
        <option value="chocolate">Chocolate</option>
        <option value="vanilla">Vanilla</option>
        <option value="strawberry">Strawberry</option>
      </select>
    </label>
  );
};
```

I have a loading component that logs every time it renders (and renders itself VERY BIG):

```ts
const Loading = () => {
  console.log("loading fallback");
  return (
    <div style={{ backgroundColor: "red", fontSize: "200px" }}>loading…</div>
  );
};
```

Finally I have a component that suspends:

```ts
const FlavorForm = () => (
  <Suspense fallback={<Loading />}>
    <FlavorPicker />
  </Suspense>
);
```

*The problem*:
Every time the user picks a flavor from the dropdown, it suspends and briefly renders the `Loading` fallback. 

How would I load the base atom async with Suspense, but then _not_ suspend when updating in memory? 

I could load in a `useEffect` and then assign the fetch result directly to the base Jotai atom. But I was hoping to do it from within Jotai. (Also I might want to reset to the original loaded value, using Jota.)

**Comment by stevemolitor** [comment]
I figured out what I did wrong; sharing in case it helps someone else. Basically I didn't listen to TypeScript. 😁 

First, in my atom setter I was accidentally merging in the new prop into a _promise_ for the value:

```ts
 (get, set, flavor: Flavor) => {
    const iceCream = get(iceCreamAtom) as IceCream;
    set(iceCreamAtom, {
      ...iceCream,
      flavor
    });
  }
```

I masked the problem via the `as IceCream` cast. Doh! The fix was to: 

a) listen to TypeScript and get rid of the cast, and 
b) make the setter an async function and properly await the value first:

```ts
 async (get, set, flavor: Flavor) => {
    const iceCream = await get(iceCreamAtom);
    set(iceCreamAtom, {
      ...iceCream,
      flavor
    });
  }
```

The flickering of the Suspense callback was caused by the fact that I had an async atom getter, and it suspended briefly while fetching:

```ts
 async (get) => {
    const { flavor } = await get(iceCreamAtom);
    return flavor;
  },
```

I found two possible fixes for that. The simple fix is just to make the getter sync:

```ts
 (get) => {
    const { flavor } = get(iceCreamAtom) as IceCream;
    return flavor;
  },
```

That gets rid of the Suspense flicker, and will work in this use case because by the time the flavor is retrieved the promise has already resolved. However it introduces a TypeScript `as` cast, and could break if I add functionality to reload the value and set to a promise again.

So a belt and suspenders solution is to use `unwrap`. Here's the entire atom definition with `unwrap`:

```ts
const flavorAtom = unwrap(
  atom(
    async (get) => {
      const { flavor } = await get(iceCreamAtom);
      return flavor;
    },
    async (get, set, flavor: Flavor) => {
      const iceCream = await get(iceCreamAtom);
      set(iceCreamAtom, {
        ...iceCream,
        flavor
      });
    }
  ),
  (prev) => prev
);
```

That gets rid of the Suspense flicker _and_ the TypeScript `as` cast, and should be safe for future "reload" refactorings. 

The `(prev) => prev` is the function that provides the default value to `unwrap` if called while suspended. I tell it to just return the previous value. This gets rid of an extra render. It works without that, but there's a duplicate render if you don't add that default value function.

Are there other solutions?

Going forward I think my personal recommendation will be to listen to TypeScript and avoid casts! While that would not have fixed the duplicate render, it would have gotten rid of the bugs. 

*Sandboxes*
- [CodeSandbox fix with sync get and typecast](https://codesandbox.io/s/elegant-bohr-khcllk?file=/src/App.tsx)
- [CodeSandbox belt and suspenders fix with unwrap](https://codesandbox.io/s/agitated-grass-hw86vv)

**Reply by dai-shi** [reply, marked as answer]
Yes, following typescript type is important. 

I think there is a better (but trickier) solution than unwrap. Basically, make the atomWithDefault type `T | Promise<T>` and you can set a non-promise value. And instead of `await` you can conditionally return a promise (this is a tricky part.)

**Comment by stevemolitor** [comment]
> I think there is a better (but trickier) solution than unwrap. Basically, make the atomWithDefault type `T | Promise<T>` and you can set a non-promise value. And instead of `await` you can conditionally return a promise (this is a tricky part.)

Ah yes that works thanks! I made a [CodeSandbox](https://codesandbox.io/s/exciting-shannon-244lmh).

I added a `isPromise` type guard function:


```ts
const isPromise = <T,>(value: unknown): value is Promise<T> =>
  typeof value === "object" &&
  value !== null &&
  typeof (value as Promise<T>).then === "function";
```

Then I changed my atom get function to conditionally return a Promise as you suggest:

```ts
  (get) => {
    const iceCream = get(iceCreamAtom);
    if (isPromise<IceCream>(iceCream)) {
      return iceCream.then(({ flavor }) => flavor);
    }
    const { flavor } = iceCream;
    return flavor;
  },
```

That works correctly, avoid duplicate rendering, and doesn't require the custom equality function that was needed with `unwrap`.

Thanks again!

**Reply by dai-shi** [reply]
Nice!
